Original notebook: 
August 2015  
Modified September 2017, October 2018

Author: FBB

Modified by: Eve Marenghi

** reproduce the result in **
http://www.mdrc.org/sites/default/files/What%20Strategies%20Work%20for%20the%20Hard%20FR.pdf 

In [1]:
import os
import sys
import numpy as np
import pylab as pl

%pylab inline

Populating the interactive namespace from numpy and matplotlib


# NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05

** $H_0: P_0 - P_1 \geq$    0 **
    
** $H_a: P_0 - P_1 $< 0    **
    
    
** $\alpha$ = 0.05    **

** this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table above): **

** $P_0=0.035, P_1=0.701$**

In [2]:

# first thing: choose your significance threshold
alpha=0.05

# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1


we must assess the statistical significance


# WHICH  TEST?

samples, categorical data
 
 ** TWO OPTIONS z test, or chi-square test.  **

# START WITH Z TEST

** the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the _assumption of normality_ **


must define the sample standard deviation (standard error)

In [3]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.4210472764645426


** standard deviation of the sampling distribution the distribution is Binomial, the binomial stdev is **

(see a proof here!: http://stats.stackexchange.com/questions/29641/standard-error-for-the-mean-of-a-sample-of-binomial-random-variables!): 

$\sqrt{\frac{p(1 - p)}{n}}$

for 2 samples this becomes 

$\sqrt{ \frac{ \hat{p}(1 - \hat{p})} {n0} + \frac{ \hat{p}(1 - \hat{p})} {n1} }$

cfr: page 138 of Statistics in a Nutshell, eq. 5.12 and here http://stattrek.com/hypothesis-test/difference-in-proportions.aspx?Tutorial=AP

** $\hat{p} =\frac{p_0  n_0 + p_1  n_1}{n_0+n_1}$ **


Note that in the online version of Statistics in a Nutshell, $\bar{x}$  should be  $\hat{p}$!!

In [4]:
# I am goonna create a little one line function to calculate the standard error
# and to calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

** z score: how many standard deviation away from the population parameter is my statistic? **

 $z=\frac{P_1-P_0}{\sigma}$

In [5]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)


20.7697865408


note that using p0-p1 or p1-p0 at the numerator is equivalent because the standardizes normal value of z has mean 0 (see image below) so that we can use the absolute value of the z score, or equivalently look for $P[Z<z]$ if z is positive, and $P[Z>z]$ if z is negative.

* if $p<\alpha$ : reject H0 **

IMPORTANT!! note that this P in the bottom line of the table is not the p-value, but 

p-value = 1-P

In [6]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


## We reject the null hypothesis because the p value (0.00) is lower than the significance level we set (alpha = 0.05). 

# what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

# NULL HYPOTHESIS: the % of former prisoners convicted of a felony in years 1-3 after release from prison is the same or higher for candidates who participated in the program as for the control group, significance level p=0.05

** $H_0: P_0 - P_1 \leq$    0 **
    
** $H_a: P_0 - P_1 $> 0    **
    
    
** $\alpha$ = 0.05    **

 $P_0 = 0.117, P_1= 0.10$

 look up data table and insert the appropriate values to get the appropriate result! you can use the functions I defined above, with different arguments. 

In [7]:
# choose the significance level 
alpha = 0.05

# 0 for control group, 1 for program group
P_0_felony = 11.7 * 0.01
P_1_felony = 10 * 0.01

if P_0_felony - P_1_felony >= 0:
    # we are done
    print ("the Null holds... but let's assess the statistical significance anyway!")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0_felony = 409
n_1_felony = 568
    
#lets get the counts by multiplying by the sample size

Nt_0_felony = P_0_felony * n_0_felony
Nt_1_felony = P_1_felony * n_1_felony

the Null holds... but let's assess the statistical significance anyway!


In [8]:
#define the sample proportion first
sp_felony = (P_0_felony * n_0_felony + P_1_felony * n_1_felony) / (n_1_felony + n_0_felony)
print("The value of the sample proportion is %2f." %sp_felony) 

The value of the sample proportion is 0.107117.


In [9]:
# One line function to figure out p and standard error 
p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1)) 

In [10]:
# Calculate the z score 
zscore = lambda p0, p1, s : (p1 - p0) / s

z_3y = zscore(P_1_felony, P_0_felony, se(p(P_0_felony, P_1_felony, n_0_felony, n_1_felony), n_0_felony, n_1_felony))
print("The z-statistic is %2f." %z_3y)

The z-statistic is 0.847640.


In [11]:
## p-value for recidivisim in 1-3 years (conviction of a felony): 
## z-statistic is 0.847640
## value read from the table for this z-statistic is 0.8023 

p_3y = 1 - 0.8023


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )

    
report_result(p_3y, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


## Since the p value (0.20) is higher than the significance level we set (alpha = 0.05), the null hypothesis holds.

# Now lets do it with the $\chi^2$ test

this analysis can also be done with the $\chi^2$ test, and the  $\chi^2$ distribution, 

see  flowcharts for how to choose a test and "Statistics In a Nutshell Chapter 4", or http://math.hws.edu/javamath/ryan/ChiSquare.html (if you are really just interested in the formula at face value)

** The chisq statistics tests the statistics calculated as : **

$\chi^2 = \sum_{i} \frac{(observation_i - expectation_i)^2}{expectation_i}$

 against a chi sq distribution.
 If we talk about sample fractions  that is 

$\chi^2 = \sum_i \frac{(f_{i,observed} - f_{i,expectated})^2}{f_{i,expected}}$

 Where _i_ indicates the sum over _each cell_.
 turns out this quantity is distributed according to a chi square distribution, so if i get the $\chi^2$ statistics i can compare it to the full chisq distribution and see how far in the tail it is

 The trickiest part (but not that tricky) is to figure out how to construct the table of values. please see Statistics In a Nutshell Chapter 4, for our data for example: Thisis called a CONTINGENCY TABLE

## ** H_0: The percentage of former prisoners who are employed after release from prison is independent of participation in the program. ** 

## ** H_1: The percentage of former prisoners who are employed after release from prison is associated with participation in the program. ** 
    
## ** $\alpha$ = 0.05    **

|                 |     success         | failure|    |               
|-----------------|:-------------------:|:-------------------:|---------------------------|
| test sample     | number of successes in test    | number of failures in test    | number members of test sample |
| control sample  | number of successes in control | number of failures in control | number members of control sample| 
|                 | total successes                |  total failures               | number of all members         |

|employed in subsadized job |     employed          | not employed     |                   
|---------------------------|:---------------------:|------------------|---------------------------|
| test sample               | $0.701*564$           | $0.299*564$      | 564                       |
| control sample            | $0.035*409$           | $0.965*4.09$     | 409                       |
|                           |                       |                  |                           |
| total                     | 409.679               |  562.912         | 973                       |

 for each cell the expected ratio is the product of the total of all rows and all columns, devided by the total

** FOR EACH CELL: **
**  expected_cell = $\frac{row~tot * col~tot}{total}$ **


In [12]:
#from evalChisq import evalChisq

In [13]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])
print(sample_values)

#print (evalChisq(sample_values))

[[ 395.364   168.636 ]
 [  12.4745  394.685 ]]


This number must be compared to the chi sq distribution. 
You must calculate the number of degrees of freedom forthis experiment. 
Generally: DOF = Number of observations - number of Independent Variables

so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

432 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

please  state what that means in terms of your Null hypothesis in a markdown cell below!

## Chi-square statistic = 432. The chi-square value is larger than 3.84. We take the value 3.84 from the Chi-Square Distribution table because we have 1 degree of freedom and our significance level is alpha = 0.05. Since 432 > 3.84, p < 0.05 (it is actually less than 0.01, since 432 > 6.63) and we **reject** the null hypothesis.

## Based on the chi-square test, there is an association between employment after release from prison and participation in the program. 

# !!!!!! TODO FOR YOU!
lets see what the chi sq statistics says about the conviction for fellonies (row 10)

* fill in the contingency table below *

|convicted of a fellony     |     yes     | no         |                   
|---------------------------|-------------|------------|----------------|
| test sample               | 0.1 * 568   |0.9 * 568   |   568          |
| control sample            |0.117 * 409  |0.883 * 409 |   409          |
|                           |             |            |                |
| total                     |  104.25     | 868.75     |   977          |

continue...

## ** H_0: The percentage of former prisoners who are convicted of a felony 1 - 3 years after being released from prison is independent of participation in the program. ** 

## ** H_1: The percentage of former prisoners who are convicted of a felony 1 - 3 years after being released from prison is associated with participation in the program. ** 
    
## ** $\alpha$ = 0.05    **

In [14]:
felony_arr = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]])
felony_tot = 0.1 * 568 + 0.9 * 568 + 0.117 * 409 + 0.883 * 409

In [15]:
# Define a function to calculate the 2x2 chi-square statistic 
def my2x2ChiSq(values, total):
    chisq = total * ((values[0,0] * values[1,1] - values[0,1] * values[1,0]) ** 2) / ((values[0,0] + values[0,1]) * (values[1,0] + values[1,1]) * (values[0,0] + values[1,0]) * (values[0,1] + values[1,1]))
    return chisq

print("The chi-square value is %2f." %my2x2ChiSq(felony_arr, felony_tot))

The chi-square value is 0.718494.


In [16]:
#DOF = Number of observcations - Number of independent variabels 
dof = 2 - 1 
print("There are %d degree(s) of freedom." %dof) 

There are 1 degree(s) of freedom.


## Chi-square = 0.72. Our chi-square value is smaller than 3.84. We take the value 3.84 from the Chi-Square Distribution table because we have 1 degree of freedom and our significance level is alpha = 0.05. Since 0.72 < 3.84, p > 0.05 (it is now between 0.25 and 0.50) and we fail to reject the null hypothesis.